### PR TITLE
String to []rune conversion

### PR DESCRIPTION
**Pull request by scriptonist**
This PR is aimed at Solving/Fixing #69, I consider this change to be aimed at getting some initial feedback. I'm new to the project and compilers in general. So as the `CONTRIBUTING.md` suggests, please be kind :sweat_smile: 

This change is heavily inspired/copied from the go source.

### CODE REVIEW

File: compiler/compiler.go [hunk 1]
```diff
@@ -2473,6 +2473,8 @@ func (c *Compiler) parseConvert(typeFrom, typeTo types.Type, value llvm.Value, p
 		switch elemType.Kind() {
 		case types.Byte:
 			return c.createRuntimeCall("stringToBytes", []llvm.Value{value}, ""), nil
+		case types.Rune:
+			return c.createRuntimeCall("stringToRunes", []llvm.Value{value}, ""), nil
 		default:
 			return llvm.Value{}, c.makeError(pos, "todo: convert from string: "+elemType.String())
```
Comment (aykevl) [inline compiler/compiler.go]: Both cases allowed by the Go spec are now supported, so it's a bug in the compiler if we get here. You can replace the `todo:` error with a panic. (Note: getting here would be a bug because when we get to SSA level the code has long been type checked and has already been verified as being valid Go).

Comment (scriptonist) [reply compiler/compiler.go]: I was really not sure, what message to put there, please help me with that.

Comment (aykevl) [reply compiler/compiler.go]: What about `unexpected type in string to slice conversion`? It's not hugely important, but it's useful if it is somewhat descriptive.

Comment (scriptonist) [reply compiler/compiler.go]: Sounds good, I've made the suggested changes, can you take a look?

File: src/runtime/string.go [hunk 2]
```diff
@@ -89,6 +89,21 @@ func stringToBytes(x _string) (slice struct {
 	return
 }
 
+// Convert a string to []rune slice.
+func stringToRunes(s string) []rune {
+	var n = 0
+	for range s {
```
Comment (deadprogram) [inline src/runtime/string.go]: Seems like it would be better to use `n := len(s)` here instead of a `range` loop, which would replace lines 94-99.

File: src/runtime/string.go [hunk 3]
```diff
@@ -89,6 +89,21 @@ func stringToBytes(x _string) (slice struct {
...
+		n++
+	}
+	var r = make([]rune, n)
+	n = 0
+	for _, e := range s {
```
Comment (deadprogram) [inline src/runtime/string.go]: Seems like you could replace the `range` here with:

```
for i := 0; i < n; i++ {
	r[i] = s[i]
}
```

**Comment by deadprogram** [comment]
Thanks for the making some changes to this PR @scriptonist I also made a few suggestions. 

Can you please add a test that uses a unicode string? Any special characters, such as this example from the golang.org web site, would be sufficient:

```
const nihongo = "日本語"
```

Thank you.

File: src/runtime/string.go (hunk 2 above)
Comment (aykevl) [reply src/runtime/string.go]: No. Iterating over a string actually decodes utf-8, while calling `len` returns the length (in bytes). What we need here are the number of Unicode code points, not the number of bytes.
See this playground for a demonstration of the difference: https://play.golang.org/p/7Nn87fbtPwU

Comment (deadprogram) [reply src/runtime/string.go]: OK great, I stand corrected.

**Comment by aykevl** [comment]
@deadprogram there is already a test included in this PR and it is actually more extensive than testing a few Chinese (?) characters because it tests various UTF-8 byte lengths (1, 2, 3, and 4 byte characters).

**Comment by scriptonist** [comment]
@deadprogram @aykevl Should I add something here, It looked like a NOP for me from the comments.

File: src/runtime/string.go (hunk 3 above)
Comment (aykevl) [reply src/runtime/string.go]: Unfortunately that doesn't work, indexing into a string returns a byte, not a rune.

**Comment by aykevl** [comment]
Thank you for this addition! I have merged the PR in e9f6e51fc8bda1824adc403494fb79ce710a006c but modified the commit message to match the style of other commit messages (we follow [Go commit message conventions](https://github.com/golang/go/wiki/CommitMessage)).

Sorry it took so long, I got really busy the last few days.
